crypto/store/loader_file.c

```diff
@@ -209,32 +213,169 @@ static OSSL_STORE_INFO *try_decode_PKCS12(const char *pem_name,
                     goto p12_end;
                 }
-                if (!PKCS12_verify_mac(p12, pass, strlen(pass))) {
-                    OSSL_STOREerr(OSSL_STORE_F_TRY_DECODE_PKCS12,
-                                  OSSL_STORE_R_ERROR_VERIFYING_PKCS12_MAC);
-                    goto p12_end;
+
+                /*
+                 * OpenSSL <1.1.0 does a naïve ISO-8859-1 to UCS-2 conversion
+                 * (i.e. treats all input characters as if they were ISO-8859-1
+                 * encoded and just zero extends them to 16 bits each) of the
+                 * pass phrase when encoding it to BMPString (and decoding).
+                 *
+                 * OpenSSL >=1.1.0 attempts a naïve UTF-8 decode (i.e. without
+                 * checking if the application locale indicates UTF-8 encoded
+                 * input) to BMPString of the pass phrase, and failing that, it
+                 * falls back on the OpenSSL <1.1.0 behaviour.
+                 *
+                 * The correct thing to do for PKCS#12 would be to convert the
+                 * pass phrase from the application's locale to a wide character
+                 * string (and hope this takes care of any normalization), and
+                 * and then simply recode that to BMPString.  That's not what
+                 * OpenSSL's PKCS#12 routines do today, though.
+                 *
+                 * We need to cater for all these possible encodings, so we
+                 * create three variants of the pass phrase:
+                 */
+
+                /*
+                 * Check variant 1: use what we get as is.
+                 * It may be perfectly legitimate UTF-8, which will be decoded
+                 * correctly, and it may be perfectly legitimate ISO-8859-1 or
+                 * ASCII, which will be correctly encded by the <1.1.0 fallback
+                 * in the majority of cases.
+                 */
+                pass = (char *)tpass;
+#ifdef OPENSSL_DEBUG_KEYGEN
+                fprintf(stderr, "Trying password as is:\n");
+                h__dump((unsigned char *)pass, strlen(pass));
+#endif
+                if (PKCS12_verify_mac(p12, pass, strlen(pass)))
+                    goto p12_parse;
+
+#if defined(__STDC_VERSION__) && __STDC_VERSION__ >= 199409L
+                /*
+                 * Check variant 2: locale to UTF-8 translation
+                 * Convert the pass phrase from the current application locale
+                 * via wide chars to UTF-8.  This should ensure that we can
+                 * read containers protected with correctly encoded pass
+                 * phrases (i.e. PKCS#12 files from other softwares as well)
+                 */
+                {
+                    wchar_t wpass[PEM_BUFSIZE];
+                    size_t wpass_len;
+                    unsigned char *pass_utf8 = NULL;
+                    size_t utf8_count, i, j;
+
+                    if ((wpass_len = mbstowcs(wpass, (char *)tpass,
```

Comment:
On Windows, does mbvstowcs convert from the code page used for file I/O, or the code page used for console I/O?

Reply:
... and which of those are you actually getting from the default PEM passphrase callback that OpenSSL provides? Both with and without the WIN32_UTF8 definition, or whatever that is...

Let's make sure that it's going to do the right thing.

Reply:
I'll have another think on that.

Reply:
mbstowcs converts according to the locale that's set by the application. https://msdn.microsoft.com/en-us/library/x99tb11d.aspx has this to say about the empty locale string:

> `setlocale( LC_ALL, "" );`
> Sets the locale to the default, which is the user-default ANSI code page obtained from the operating system.

The default PEM callback gives exactly what it got from the built-in UI method, i.e. the exact characters it got fed, except on Windows with OPENSSL_WIN32_UTF8 defined, where whatever was given by the user is converted to UTF-8

Reply:
> the exact characters it got fed, except on Windows with OPENSSL_WIN32_UTF8 defined,

Let's say I have US locale, Greek keyboard and *don't* set OPENSSL_WIN32_UTF8. What does application get? Just a bunch of question marks actually. Are these exact characters it got fed? Well, one can argue about the feeding point, but Windows kernel did pass Greek and it was user-land in the current process that converted them to question marks, just not libcrypto. Libcrypto simply made wrong calls...

Reply:
Re the notorious 0xC3 0xAF, in *variant 2*, it will get converted to wchar_t U+00C3 U+00AF if the active code page allows it, or I assume U+00EF if the code page happens to be 65001.  However, should 0xC3 0xAF be an exact sequence that *shouldn't* be interpreted as a UTF-8 character, variant 2 will fail and variant 3 will hopefully succeed.

As for greek keyboard and US locale, are you saying that you'd get something other than question marks for the greek characters if `setlocale(LC_ALL, "")` was avoided?  I assume this is something that happens in `ui_openssl.c`, right?  Or?

Reply:
Once again, this code is not even compiled on Windows. If it was, setlocate(LC_ALL,"") will set current **ANSI** code page, and depending on it, 0xC3 0xAF can be translated to anything, for example U+0102 U+17b on Hungarian system. or U+0393 U+2015 on Greek. It will *always* get translated. Not calling setlocale will yield U+00C3 U+00AF though. This last part is different from Linux, as if you don't setlocal, mbstowcs will fail. It will never get translated to U+00EF. Well, unless you call setlocale(LC_ALL,"some_thing"), i.e. with explicit argument, and "some_thing" has to trigger UTF-8. For example "Hungarian_Hungary" wouldn't. ["hu_HU" would.]

As for remark about Greek keyboard. It's really about non-overlapping alphabets. I don't think that call to setlocale would have **any** effect. Because setlocale is pure MSVCRT thing, while console is read with Win32 call, non-MSVCRT thing, ReadConsoleA. ReadConsoleA is essentially ReadConsoleW followed by WideCharToMulitByte **with** CP_ACP. There is no translation from Greek wide char to US narrow char, and so it will be inevitably replaced with question mark.

Reply:
Consider following. You have an 0xC3 0xAF string that you picked from a file, i.e. didn't type. It's in Hungarian, and ĂŻ is correct PKCS#12 password. Now consider what's happening with it here, in this case. On Windows without calling setlocale, mbstowcs's output is U+00C3, U+00AF. No good. Will 3rd attempt catch it? No. On [Hungarian] Windows with calling setlocale. Output is U+0102 U+017b. Correct password. Now on Linux without calling setlocale. mbstowcs fails, Will 3rd attempt catch it? No. On Hungarian Linux with calling setlocale. Note that on contemporary Hungarian system default locale would be hu_HU.utf8. Output is U+00EF, it was already tried by 1st try. It's wrong password, 3rd attempt won't catch it. In order to make it work you have to set LC_TYPE to hu_HU, i.e. without utf8 suffix, and call setlocale. Only them output is correct. In the essence that is what I meant by initial remark about calling vs. not calling setlocale. I mean calling or not calling is matter of succeeding or not succeeding. Right thing to do is to convert 0xC3 0xAF to UTF-8 by *external* means and use that relying on initial PCKS#12 to succeed.

Reply:
Now let's say you *type* ĂŻ on Hungarian keyboard and it's correct PKCS#12 password. On Windows if you set OPENSSL_WIN32_UTF8 environment variable, first PKCS#12 call succeeds irregardless whether or not application called setlocale, and irregardless whether or not your Windows system locale is Hungarian. It just works. If you don't set OPENSSL_WIN32_UTF8, second calls succeeds *only* on Hungarian Windows and *only* if application called setlocale. [Well, *if* code is compiled even on Windows.] If you type it on *any* Linux with UTF-8 locale first call to PCKS#12 succeeds. If you go the mile and set hu_HU, i.e. without utf8 suffix, second call to PKCS#12 succeeds if application called setlocale and fails if it didn't.

Reply:
I'm sorry, I fail to see why a string translated to wide chars according to the current code page would be a bad thing.  If someone types in "ГЇ" and the system code page is 1251, we will get 0xC3 0xAF in `tpass`, will we not?  Why should it be translated to U+00EF when neither the input method nor the code page indicates UTF-8?

Of course, the application can set whatever locale they like, but mustn't they ensure that whatever they pass as strings are encoded accordingly?

Also, please enlighten me...  among all the calls you mentioned, I'm not sure I understood which one would generate question marks for Greek characters.  And I don't quite understand where `CP_ACP` comes in, 'cause all I can see in `apps/win32_init.c` and `crypto/ui/ui_openssl.c` is `CP_UTF8`.  The only places I can see `CP_ACP` used is with file names, which I assume is appropriate, but you know more about that than I do...

Reply:
But that aside, if you're looking to tune things up for Windows, what would *you* do?

Reply:
More comments, ok hang on...

Reply:
> I fail to see why a string translated to wide chars according to the current code page would be a bad thing.

I didn't say that it's a bad thing. I merely pointing out that there is dependency on whether or not application called setlocale(LC_ALL,"").

Reply:
> among all the calls you mentioned, I'm not sure I understood which one would generate question marks for Greek characters

ReadConsoleA.

Reply:
> And I don't quite understand where CP_ACP comes in

That's how ReadConsoleA is implemented. You call ReadConsoleA, ReadConsoleA calls ReadConsoleW and then WideByteToMutliChar with CP_ACP. You have no control over it.

Reply:
> If someone types in "ГЇ" and the system code page is 1251, we will get 0xC3 0xAF in tpass, will we not?

Without setting OPENSSL_WIN32_UTF8, yes. Next question is what is correct password. If you had Windows generate the target PCKS#12 object, then correct password is the one tpass **would** hold if you called openssl **with** OPENSSL_WIN32_UTF8 set. Now, whether or not you hit correct password **without** OPENSSL_WIN32_UTF8 depends on whether or not application called setlocale(LC_ALL,"") (provided that code is compiled on Windows). Yes, you suggest in this request that openssl binary does it. But us providing library implies that it be linked with an application with have no control over...

Reply:
> I didn't say that it's a bad thing. I merely pointing out that there is dependency on whether or not application called `setlocale(LC_ALL,"")`.

Correct.  Not calling it is equivalent to `setlocale(LC_ALL, "C")`, and with that, `mbstowcs` fails on anything non-ASCII, at least on Linux (I assume the same goes for Windows).  Calling it with the empty string sets the locale to whatever your system default is (according to environment variables on Unix, according to some code page setting on Windows), and `mbstowcs` will translate the string to wide chars if it can.

I guess the biggest problem on Windows is really with `ui_openssl.c`, which either returns UTF8 (if `OPENSSL_WIN32_UTF8` is defined) without having the code page 65001, or potentially strings with certain characters changed to question marks in the ReadConsoleA case (although quite frankly, that sounds like a faulty configuration, just like an input method that generates ISO-8859-2 encoded characters would become problematic if the locale indicated ISO-8859-1).

This is possibly a case where we should check for `OPENSSL_WIN32_UTF8` here as well, and simply assume that we can skip over the `mbstowcs` call and following wide to UTF-8 conversion in that case...  does that sound right to you?  (also assuming we find the right checks to get this code compiled on Windows to begin with)

Reply:
> This is possibly a case where we should check for `OPENSSL_WIN32_UTF8` here as well, and simply assume that we can skip over the `mbstowcs` call and following wide to UTF-8 conversion in that case...  does that sound right to you?  (also assuming we find the right checks to get this code compiled on Windows to begin with)

I'm sorry, this suggestion is silly, it's simply doing nothing about the pass phrase, which has already been covered by variant 1

Reply:
> Now, whether or not you hit correct password **without** OPENSSL_WIN32_UTF8 depends on whether or not application called setlocale(LC_ALL,"") (provided that code is compiled on Windows). Yes, you suggest in this request that openssl binary does it. But us providing library implies that it be linked with an application with have no control over...

Yes, it's true that this requires the cooperation of the application in that manner.  Do you have a suggestion as to what to do about it?  Should we `setlocale(LC_ALL, "")` internally?

Reply:
(I'm off to be, btw...  be back tomorrow)

Reply:
> Not calling it is equivalent to setlocale(LC_ALL, "C"), and with that, mbstowcs fails on anything non-ASCII, at least on Linux (I assume the same goes for Windows).

No. See above. Windows is happy to mbstowcs with "C" locale.

> mbstowcs will translate the string to wide chars if it can.

Again, see above, Windows mbstowcs will *always* translate to something. Because setlocale(LC_ALL,"") always set "ANSI" locale.

> without having the code page 65001

65001 is not a code page you can set as system one. There is no code page you can set that would make setlocale(LC_ALL,"") make mbstowcs interpret input as UTF-8. Though I might be wrong about this. It might happen that Asian code pages could... You can call setlocale(LC_ALL,"some-non-empty-string") that can make mbstowcs interpret input as UTF-8, but as long as it's setlocale(LC_ALL,""), it's never UTF-8 [Again, I'm not sure how things are on Asian Windows versions.]